### PR TITLE
Fixed issue with robots going into forbidden areas when stealing/receiving

### PR DESCRIPTION
**Pull request by jvogt23**
## Description
Describe your pull request.
Title.
## Associated / Resolved Issue
Resolves # or [ClickUp card](link-to-clickup-card)
86ayq6wkp
## Design Documents
[Link](link-to-design-doc)

### Test Case 1: Ball going into goal area 
1. Step 1: Start a match and set the ball with velocity going through the goal area
2. Step 2: Observe the robots going out of stealing/receiving mode and into seeking mode. Should happen whenever the ball is in the red
3. Step 3: When the ball leaves the red, they should continue chasing the ball

**Expected result:**See steps 2 and 3

## Key Files to Review
Group 1: Position
 * File 1: Offense.cpp, offense.hpp
 * File 2

Group 2: planning
 * File 3: collect_path_planner
 * File 4

## Review Checklist

- [ ] **Docstrings**: All methods and classes should have the file appropriate docstrings which follow the guidelines in the ["Contributing" page](https://rj-rc-software.readthedocs.io/en/latest/contributing.html) of our docs.
- [ ] **Remove extra print statements**: Any print statements used for debugging should be removed
- [ ] **Tag reviewers**: Tag some people for review and ping them on Slack

## (Optional) Sub-issues (for drafts)
_Note: if you find yourself breaking this PR into many smaller features, it may make sense to break up the PR into logical units based on these features._
- [ ] Step 1
- [ ] Step 2

### CODE REVIEW

File: soccer/src/soccer/strategy/agent/position/offense.cpp [hunk 1]
```diff
@@ -559,4 +566,15 @@ rj_geometry::Point Offense::calculate_best_shot() const {
     }
     return best_shot;
 }
+
+// Checks whether ball is out of range for stealing/receiving
+bool Offense::ball_in_red(WorldState* last_world_state_) {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/offense.cpp]: do not pass in `last_world_state_` as a function parameter; you are simply hiding the class-scope `last_world_state_` that already exists. use this directly. in other words, you can just remove this parameter and your function will work perfectly.

Comment (sid-parikh) [reply soccer/src/soccer/strategy/agent/position/offense.cpp]: can you incorporate this function into `should_steal`?

File: soccer/src/soccer/strategy/agent/position/offense.cpp [hunk 2]
```diff
@@ -37,7 +36,7 @@ Offense::State Offense::next_state() {
 
         case SEEKING: {
             // If the ball seems "stealable", we should switch to STEALING
-            if (can_steal_ball()) {
+            if (can_steal_ball() && !ball_in_red(last_world_state_)) {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/offense.cpp]: incorporate the `ball_in_red` into `can_steal_ball()` they can be separate functions but have `can_steal_ball` check `ball_in_red` directly

File: soccer/src/soccer/strategy/agent/position/offense.cpp [hunk 3]
```diff
@@ -131,6 +134,10 @@ Offense::State Offense::next_state() {
                 return POSSESSION_START;
             }
 
+            if (ball_in_red(last_world_state_)) {
+                return SEEKING_START;
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/offense.cpp]: return `DEFAULT` instead

File: soccer/src/soccer/strategy/agent/position/offense.cpp [hunk 4]
```diff
@@ -559,4 +566,15 @@ rj_geometry::Point Offense::calculate_best_shot() const {
...
     return best_shot;
 }
+
+// Checks whether ball is out of range for stealing/receiving
+bool Offense::ball_in_red(WorldState* last_world_state_) {
+    auto& ball_pos = last_world_state_->ball.position;
+    if (field_dimensions_.our_defense_area().contains_point(ball_pos) ||
+        field_dimensions_.their_defense_area().contains_point(ball_pos) ||
+        !field_dimensions_.field_rect().contains_point(ball_pos)) {
+        return true;
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/offense.cpp]: GH actions (clang-tidy) has left a warning here. you should return the boolean literal directly instead.

File: soccer/src/soccer/planning/planner/collect_path_planner.cpp [hunk 5]
```diff
@@ -270,8 +275,19 @@ Trajectory CollectPathPlanner::fine_approach(
     motion_constraints_hit.max_speed =
         std::min(target_hit_vel.mag(), motion_constraints_hit.max_speed);
 
-    Trajectory path_hit = CreatePath::simple(start_instant.linear_motion(), target_hit,
-                                             plan_request.constraints.mot, start_instant.stamp);
+    // Trajectory path_hit = CreatePath::simple(start_instant.linear_motion(), target_hit,
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/collect_path_planner.cpp]: please remove this commented-out code

File: soccer/src/soccer/planning/planner/collect_path_planner.cpp [hunk 6]
```diff
@@ -358,8 +374,18 @@ Trajectory CollectPathPlanner::control(const PlanRequest& plan_request, RobotIns
     // Try to use the RRTPathPlanner to generate the path first
     // It reaches the target better for some reason
     std::vector<Point> start_end_points{start.position(), target.position};
-    Trajectory path = CreatePath::rrt(start.linear_motion(), target, motion_constraints,
-                                      start.stamp, static_obstacles, dynamic_obstacles);
+
+    // Trajectory path = CreatePath::rrt(start.linear_motion(), target, motion_constraints,
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/collect_path_planner.cpp]: here as well

File: soccer/src/soccer/planning/planner/collect_path_planner.cpp [hunk 7]
```diff
@@ -110,6 +110,11 @@ Trajectory CollectPathPlanner::plan(const PlanRequest& plan_request) {
     std::vector<DynamicObstacle> dynamic_obstacles;
     fill_obstacles(plan_request, &static_obstacles, &dynamic_obstacles, false);
 
+    if (static_obstacles.hit(ball.position) ||
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/collect_path_planner.cpp:116]: please add a comment here explaining our reasoning

**Review by sid-parikh — CHANGES_REQUESTED** [review]
functionality is what we want, thanks. just some code quality nits.

File: soccer/src/soccer/strategy/agent/position/offense.cpp (hunk 1 above)
Comment (jvogt23) [reply soccer/src/soccer/strategy/agent/position/offense.cpp]: Do you mean the function can_steal_ball() in Offense?

**Comment by sid-parikh** [comment]
@jvogt23 apologies but please rebase with the latest version of offense

**Review by sid-parikh — APPROVED** [review]

**Pull request merged**